### PR TITLE
JM - implement image tag_list editing

### LINKED ISSUE #14: Modify Tag List
#### As a user I want to edit the tag list for an image.

__Story__:

Image tags are awesome -- our user's love them! Tags help us make sense of the
plethora of images added to our system. But how annoying is it when our users
make spelling mistakes in their image tags? Super annoying! How awesome, and
useful would it be to allow for tags to be edited after an image is added?
Super awesome, and useful!

__Acceptance criteria__:
- [x] There is a link to edit the tags for an image.
- [x] The image whose tags are to be modified is displayed on the edit page.
- [x] On the edit page, I can change the tags to any valid tag list.
- [x] After saving the edit page with valid tags, my changes are visible
  throughout the rest of the application.

__Discussion__:
- [ ] Should an image's URL be editable?

__Dependencies__:
- Image Tags
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by jmangel**
Closes #14

### CODE REVIEW

File: app/assets/stylesheets/_show-page-image.scss [hunk 1]
```diff
@@ -1,4 +1,4 @@
 .show-page-image {
-  max-height: 100%;
-  max-width: 100%;
+  max-height: 80vh;
```
Comment (wilsonyqm) [inline app/assets/stylesheets/_show-page-image.scss]: is it related to this story? If it is can you make it a separate commit? If not, just remove them, or create another story for it.

Comment (jmangel) [reply app/assets/stylesheets/_show-page-image.scss]: It's not related to the story, I can definitely move it into another commit, but I would rather not make a whole pull request for it

Comment (jmangel) [reply app/assets/stylesheets/_show-page-image.scss]: Although I can, it's not priority to get it merged so I can submit it to Aaron for review

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -26,6 +26,22 @@ def create
...
+  end
+
+  def update
+    if @image.update_attributes(image_update_params)
+      flash[:success] = 'Image successfully updated'
+      return redirect_to @image
+    end
+    errors = @image.errors
+    load_image
+    errors.messages.each do |key, value|
```
Comment (wilsonyqm) [inline app/controllers/images_controller.rb]: what is this line for?

Comment (jmangel) [reply app/controllers/images_controller.rb]: This was an implementation decision. I wanted to reload the original image data (ie. tags) into the form with `load_image` in the case of invalid update, and the lines about errors are to display the errors that otherwise would be cleared by reloading `@image`.

File: app/views/images/_image_info.html.erb [hunk 3]
```diff
@@ -1,5 +1,6 @@
 <%= render partial: 'tags', object: image.tag_list %>
 <div class="d-flex <%= local_assigns.fetch :buttons_class, '' %>">
   <%= button_to 'Delete', image, class: 'btn btn-danger', form_class: 'mr-1', method: :delete, data: { confirm: 'Are you sure?' } %>
-  <%= button_to 'Share', share_image_path(image), class: 'btn btn-secondary js-share-image', data: { image_id: image.id } %>
+  <%= button_to 'Share', share_image_path(image), class: 'btn btn-secondary js-share-image', form_class: 'mr-1', data: { image_id: image.id } %>
+  <%= button_to 'Edit', edit_image_path(image), class: 'btn btn-secondary', method: :get, data: { image_id: image.id } %>
```
Comment (wilsonyqm) [inline app/views/images/_image_info.html.erb]: I believe `method: get` is by default.

Comment (wilsonyqm) [reply app/views/images/_image_info.html.erb]: What is the image_id here used for?

Comment (jmangel) [reply app/views/images/_image_info.html.erb]: I think for `button_to` the default is `:post`. For `link_to` I believe you're right

File: app/views/images/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<div class="jumbotron jumbotron-fluid <%= add_jumbotron_offset_classes(@images) %>">
```
Comment (wilsonyqm) [inline app/views/images/edit.html.erb]: can you explain what is an `add_jumbotron_offset_classes` here?

File: app/views/images/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  <div class="container">
+    <h1 class="display-3">Edit image</h1>
```
Comment (wilsonyqm) [inline app/views/images/edit.html.erb]: what is `display-3` for?

File: app/views/images/_image_info.html.erb (hunk 3 above)
Comment (jmangel) [reply app/views/images/_image_info.html.erb]: Oh, I will remove `image_id` from the `edit` button

File: app/views/images/edit.html.erb (hunk 4 above)
Comment (jmangel) [reply app/views/images/edit.html.erb]: Yes, it's a helper method that adds a styling class that's dependent on whether or not there are images in the image grid on the page. It makes the jumbotron background a little taller to create the overlapping style of the image grid on top of the jumbotron

File: app/views/images/edit.html.erb (hunk 5 above)
Comment (jmangel) [reply app/views/images/edit.html.erb]: It's a bootstrap typography class https://v4-alpha.getbootstrap.com/content/typography/#display-headings

File: app/views/images/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    <%= simple_form_for @image, url: image_url(@image) do |f| %>
+        <%= f.input :tag_list, input_html: { value: f.object.tag_list.join(", ") }, placeholder: 'Tags (separate with commas)', class: 'form-control', label: 'Tags (separate with commas)', label_html: { class: 'sr-only' } %>
```
Comment (wilsonyqm) [inline app/views/images/edit.html.erb]: what is `sr-only`. One recommendation for your naming: please make it clear and try to make it complete.

Comment (jmangel) [reply app/views/images/edit.html.erb]: It's a bootstrap class for accessibility. `Screen readers will have trouble with your forms if you don't include a label for every input. For these inline forms, you can hide the labels using the .sr-only class.`
https://getbootstrap.com/docs/3.3/css/

File: app/views/images/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+        <div class="form-group">
+          <%= f.button :submit, 'Update' %>
+        </div>
+    <% end %>
+  </div>
+</div>
+
+<%= image_tag(@image.url, class: 'half-page-image') %>
```
Comment (wilsonyqm) [inline app/views/images/edit.html.erb]: why here shows the image_tag again when it's already in the form.

File: app/views/images/edit.html.erb (hunk 6 above)
Comment (wilsonyqm) [reply app/views/images/edit.html.erb]: Ohh, well, that's interesting to know.

File: config/routes.rb [hunk 8]
```diff
@@ -5,7 +5,7 @@
 
   get '/images/filter/:tag', to: 'images#index'
 
-  resources :images, only: [:index, :create, :show, :destroy] do
+  resources :images, only: [:index, :create, :show, :destroy, :edit, :update] do
```
Comment (wilsonyqm) [inline config/routes.rb]: we don't need the only here anymore

File: app/assets/stylesheets/_show-page-image.scss (hunk 1 above)
Comment (wilsonyqm) [reply app/assets/stylesheets/_show-page-image.scss]: I would recommend a separate PR.

File: config/routes.rb (hunk 8 above)
Comment (jmangel) [reply config/routes.rb]: Good point, should I use `except: :new` though?

File: app/controllers/images_controller.rb (hunk 2 above)
Comment (wilsonyqm) [reply app/controllers/images_controller.rb]: you can do @image.reload!, but I don't think it's necessary

File: config/routes.rb (hunk 8 above)
Comment (wilsonyqm) [reply config/routes.rb]: Sure

File: test/controllers/images_controller_test.rb [hunk 9]
```diff
@@ -67,6 +67,37 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+                           tag_list: tag_list)
+
+    ImageUrlValidator.any_instance.expects(:validate_each).returns(true)
+    put image_url(@image), params: { image: {
+      tag_list: ''
+    } }
+
+    assert_response :unprocessable_entity
+    @image.reload
+    assert_equal tag_list, @image.tag_list
```
Comment (wilsonyqm) [inline test/controllers/images_controller_test.rb]: can you assert the error message too?

File: test/controllers/images_controller_test.rb [hunk 10]
```diff
@@ -67,6 +67,37 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+  end
+
+  test 'unsuccessful update' do
+    ImageUrlValidator.any_instance.expects(:validate_each).returns(true)
+    tag_list = ['test', 'Jeff Goldblum']
+    @image = Image.create!(url: 'https://pbs.twimg.com/media/CK2KzPfWwAAOn36.jpg',
+                           tag_list: tag_list)
+
+    ImageUrlValidator.any_instance.expects(:validate_each).returns(true)
+    put image_url(@image), params: { image: {
```
Comment (wilsonyqm) [inline test/controllers/images_controller_test.rb]: no need for put.

File: test/flows/images_crud_test.rb [hunk 11]
```diff
@@ -26,6 +26,34 @@ class ImagesCrudTest < FlowTestCase
...
+    Image.create!(url: cute_puppy_url, tag_list: 'puppy, cute')
+
+    images_index_page = PageObjects::Images::IndexPage.visit
+    assert_equal 1, images_index_page.images.count
+    assert images_index_page.showing_image?(url: cute_puppy_url)
+
+    image_to_edit = images_index_page.images.find do |image|
+      image.url == cute_puppy_url
+    end
+    image_edit_page = image_to_edit.edit!
```
Comment (wilsonyqm) [inline test/flows/images_crud_test.rb]: can you test the invalid case here as well?

Comment (jmangel) [reply test/flows/images_crud_test.rb]: 👍
